### PR TITLE
[Testcase Refactoring] Device-agnostic move_to_device_pass tests via instantiate + device

### PR DESCRIPTION
**Pull request by SHIHAIFENG15**
## Summary

Make `move_to_device_pass` tests in `test_passes.py` device-agnostic via `instantiate_device_type_tests` and injected `device`, so accelerators including privateuse1 (NPU) are covered.

## Changes

- Move `move_to_device` cases into `TestPassesDevice` with `hw_classification = ACCELERATOR` and `except_for="cpu"`.
- Replace hardcoded `"cuda"` / `@requires_gpu` with the injected `device`.
- Keep `assertExpectedInline` goldens, parameterized by `device_type`.
- Keep the RNN case in the same `TestPassesDevice` class (no separate `TestPassesRNN`).

## Test Plan

```bash
python test/export/test_passes.py -v -k TestPassesDevice
```
Local NPU:

python test/export/test_passes.py -v -k TestPassesDevice
(with PYTORCH_TESTING_DEVICE_ONLY_FOR=privateuse1, import torch_npu before run)

Result: Ran 4 tests — 3 passed, 1 error

PASS: test_move_device_to_npu, test_move_device_submod_npu, test_move_device_example_inputs_npu
ERROR: test_move_to_device_pass_npu — DynamicGRUV2 unsupported on NPU (float32 GRU); torch_npu backend gap, not a pass regression
This is a torch_npu backend gap (DynamicGRUV2), not a move_to_device_pass bug. The case stays on the accelerator class (no per-device skip). Tracking: https://gitcode.com/Ascend/pytorch/issues/3886
CPU: not run (TestPassesDevice uses except_for="cpu").

@wjlFlyer @pjyFight @FuDdd @fffrog @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

### CODE REVIEW

File: test/export/test_passes.py [hunk 1]
```diff
@@ -1324,51 +1327,99 @@ def forward(self, x):
...
+    to = torch.ops.aten.to.dtype_layout(arg0_1, dtype = torch.float32, layout = torch.strided, device = '{device_type}');  arg0_1 = None
     add = torch.ops.aten.add.Tensor(to, to);  to = None
     return (add,)
     """,
         )
 
-    @unittest.skipIf(not TEST_CUDA, "requires cuda")
-    def test_move_to_device_pass(self):
+    def test_move_device_example_inputs(self, device):
+        device_type = str(device).split(":", 1)[0]
```
Comment (can-gaa-hou) [inline test/export/test_passes.py]:
```suggestion
        device_type = torch.device(device).type
```

File: test/export/test_passes.py [hunk 2]
```diff
@@ -1324,51 +1327,99 @@ def forward(self, x):
...
+        location = torch.device(device0)
+        ep_moved = move_to_device_pass(ep, location=location)
+
+        # Verify example_inputs moved
+        self.assertEqual(ep_moved.example_inputs[0][0].device, torch.device(device0))
+        self.assertEqual(ep_moved.example_inputs[0][1].device, torch.device(device0))
+        self.assertEqual(ep_moved.example_inputs[1]["z"].device, torch.device(device0))
+
+    def test_move_to_device_pass(self, device):
+        device_type = str(device).split(":", 1)[0]
```
Comment (can-gaa-hou) [inline test/export/test_passes.py]: Ditto

File: test/export/test_passes.py [hunk 3]
```diff
@@ -1324,51 +1327,99 @@ def forward(self, x):
...
+        if device_type not in ("cuda", "xpu"):
+            self.skipTest("move_to_device_pass RNN is supported only on cuda/xpu")
```
Comment (can-gaa-hou) [inline test/export/test_passes.py]: If this test is not device-generic, please add another device-specific class (e.g. TestPassesCUDA or something?) and move to it. Try your best not to explicitly skip a specific device in the test unless you have to.

**Review by can-gaa-hou — COMMENTED** [review]
Thanks for the PR! Overall lgtm minor some small nits. Please take a look.

File: test/export/test_passes.py (hunk 3 above)
Comment (SHIHAIFENG15) [reply test/export/test_passes.py]: Thanks for the review. Addressed the nits:
- use `torch.device(device).type` instead of string split
- move the RNN case into `TestPassesRNN` with `only_for=("cuda", "xpu")` (no in-method device skip)

File: test/export/test_passes.py (hunk 2 above)
Comment (SHIHAIFENG15) [reply test/export/test_passes.py]: done

File: test/export/test_passes.py (hunk 1 above)
Comment (SHIHAIFENG15) [reply test/export/test_passes.py]: > 

done

File: test/export/test_passes.py [hunk 4]
```diff
@@ -1378,62 +1433,34 @@ def forward(self, x):
...
-        # Move to CUDA
-        location = torch.device("cuda:0")
-        ep_cuda = move_to_device_pass(ep, location=location)
 
-        # Verify example_inputs moved to CUDA
-        self.assertEqual(ep_cuda.example_inputs[0][0].device, torch.device("cuda:0"))
-        self.assertEqual(ep_cuda.example_inputs[0][1].device, torch.device("cuda:0"))
-        self.assertEqual(ep_cuda.example_inputs[1]["z"].device, torch.device("cuda:0"))
+instantiate_device_type_tests(TestPassesDevice, globals(), except_for="cpu")
+instantiate_device_type_tests(TestPassesRNN, globals(), except_for="cpu")
```
Comment (lizhihai137) [inline test/export/test_passes.py]: Why is it necessary to split the TestPassesRNN test case when the accelerator types of the test cases are the same?

**Review by lizhihai137 — CHANGES_REQUESTED** [review]
please review

File: test/export/test_passes.py (hunk 4 above)
Comment (SHIHAIFENG15) [reply test/export/test_passes.py]: Thanks — merged TestPassesRNN into TestPassesDevice since both used the same ACCELERATOR classification and except_for="cpu" instantiate settings.

**Review by lizhihai137 — APPROVED** [review]
thank you for your contributions

File: test/export/test_passes.py [hunk 5]
```diff
@@ -1324,51 +1327,97 @@ def forward(self, x):
...
-    to = torch.ops.aten.to.dtype_layout(arg0_1, dtype = torch.float32, layout = torch.strided, device = 'cuda');  arg0_1 = None
+    _assert_tensor_metadata_default = torch.ops.aten._assert_tensor_metadata.default(arg0_1, dtype = torch.float32, device = '{device_type}', layout = torch.strided);  _assert_tensor_metadata_default = None
+    to = torch.ops.aten.to.dtype_layout(arg0_1, dtype = torch.float32, layout = torch.strided, device = '{device_type}');  arg0_1 = None
     add = torch.ops.aten.add.Tensor(to, to);  to = None
     return (add,)
     """,
         )
 
-    @unittest.skipIf(not TEST_CUDA, "requires cuda")
-    def test_move_to_device_pass(self):
```
Comment (fffrog) [inline test/export/test_passes.py:1371]: Why swap the positions of `test_move_to_device_pass` and `test_move_device_example_inputs`? This would increase the difficulty of review.

**Review by fffrog — DISMISSED** [review]
Overall look ok to me ,thank you for your contributions.

File: test/export/test_passes.py (hunk 5 above)
Comment (SHIHAIFENG15) [reply test/export/test_passes.py:1371]: The method order follows the original TestPasses class in main (move_device_example_inputs → move_device_submod → move_device_to → move_to_device_pass).

**Comment by SHIHAIFENG15** [comment]
Updated the PR to preserve the complete device injected by `instantiate_device_type_tests` instead of reconstructing `<device_type>:0`. Device-type-only values are still used where required, such as autocast and the type-only remapping checks. I also enabled XPU instantiation with `allow_xpu=True`. The RNN test remains device-agnostic with the other accelerator tests.

**Review by lizhihai137 — APPROVED** [review]

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

**Comment by SHIHAIFENG15** [comment]
> **Claude finished @can-gaa-hou's task in 14m 17s** —— [View job](https://github.com/pytorch/pytorch/actions/runs/33827507068)
> 
> Reviewed with the `/pr-review` skill — the refactor is directionally right, but the f-string `assertExpectedInline` goldens break the `EXPECTTEST_ACCEPT=1` workflow and the newly-enabled XPU path appears unverified.
> 
> Full PR review (click to expand)
> ## PR Review: #191079
> ### Summary
> Converts the four `move_to_device_pass` tests in `test/export/test_passes.py` from hardcoded-`cuda` + `skipIf(not TEST_CUDA)` into a `TestPassesDevice` class driven by `instantiate_device_type_tests(..., except_for="cpu", allow_xpu=True)`. The device plumbing itself is correct, but the golden strings were converted to f-strings, which breaks expecttest's accept path, and the test plan shows no CUDA or XPU run despite the PR newly enabling XPU.
> 
> ### Testing
> **1. `f"""..."""` goldens break the `EXPECTTEST_ACCEPT=1` accept workflow.** `test/export/test_passes.py:1349` and `:1372`
> 
> ```python
> self.assertExpectedInline(
>     ep.graph_module.code.strip("\n"),
>     f"""\
> def forward(self, x):
>     _assert_tensor_metadata_default = ... device = '{device_type}' ...
> ```
> 
> expecttest's accept path rewrites the _source literal_ in place. Two problems follow:
> 
> * Its literal-matching regex handles only the `r` prefix (`(?<=\W)([+~]?)(r?)(["'])...`); the `f` prefix means the character preceding `"""` is `\w`, so the lookbehind fails at the intended position. The repo already documents this class of failure: `test/dynamo/test_export.py:3483` carries `# TODO: This is naughty, EXPECTTEST_ACCEPT=1 doesn't work` on a `.format()`-parameterized golden.
> * Even if it did match, accepting would bake the concrete runtime device (`'cuda'`) into the source and destroy the `{device_type}` placeholder — silently converting the test back to CUDA-only.
> 
> These are 30-line multi-line graph goldens; `--accept` is the primary way they get maintained when the exporter's printed IR changes. Losing it is a real maintainability regression, not a cosmetic one.
> 
> The established pattern is a **static** golden plus post-hoc normalization — see `test/inductor/test_max_autotune.py:2422-2438`:
> 
> ```python
> expected = """{
>     'input_nodes':[
>         "[[10,22],[22,1],torch.float32,device(type='cuda',index=0),0]", ...
> ```
> 
> ```python
> expected = expected.replace("cuda", GPU_TYPE)
> self.assertExpectedInline(remove_white_space(cache_key), remove_white_space(expected))
> ```
> 
> Applying that here keeps the literal static (so `--accept` works) while still parameterizing the device. [Fix this →](https://claude.ai/code?q=In%20test%2Fexport%2Ftest_passes.py%20on%20branch%20npu-export-device-agnostic-passes%2C%20tests%20test_move_device_to%20(line%20~1336)%20and%20test_move_device_submod%20(line%20~1358)%20use%20f-string%20goldens%20passed%20to%20assertExpectedInline.%20This%20breaks%20the%20EXPECTTEST_ACCEPT%3D1%20accept%20workflow%20because%20expecttest%20rewrites%20the%20source%20literal%20in%20place%20and%20its%20literal-matching%20regex%20only%20handles%20the%20r%20prefix%2C%20not%20f.%20Refactor%20both%20to%20use%20a%20STATIC%20triple-quoted%20golden%20containing%20the%20literal%20string%20cuda%20as%20the%20placeholder%2C%20then%20call%20.replace(%22cuda%22%2C%20device_type)%20on%20the%20expected%20string%20before%20passing%20it%20to%20assertExpectedInline.%20Follow%20the%20existing%20pattern%20at%20test%2Finductor%2Ftest_max_autotune.py%20line%202434.%20Do%20not%20change%20any%20other%20behavior.&repo=pytorch/pytorch)
> 
> **2. `allow_xpu=True` newly enables these goldens on XPU, but the test plan shows no XPU or CUDA run.** `test/export/test_passes.py:1450-1452`
> 
> `allow_xpu=True` is what causes `XPUTestBase` to be appended (`common_device_type.py:1035-1036`) — XPU is _not_ in the default bases (`get_device_type_test_bases`, `common_device_type.py:931-936`). So this PR turns on a device that was never exercised by these tests before, and the PR body reports only a privateuse1/NPU run ("CPU: not run", no CUDA, no XPU). Since two of the four tests assert on exact printed IR and a third runs `torch.nn.GRU` on-device, "should work" is not sufficient — please post the output of:
> 
> ```shell
> python test/export/test_passes.py -v -k TestPassesDeviceXPU
> python test/export/test_passes.py -v -k TestPassesDeviceCUDA
> ```
> 
> **3. The GRU test is known-failing on the exact backend this PR exists to enable, with no skip and no tracking issue.** `test/export/test_passes.py:1413-1447`
> 
> The PR body states `test_move_to_device_pass_npu` **errors** ("DynamicGRUV2 unsupported on NPU"). So after this lands, privateuse1 gets a permanently red test with nothing in-tree explaining it. `@can-gaa-hou` asked earlier to avoid per-device skips, and that is the right instinct — but the alternative cannot be shipping a known failure silently. Please either link a `torch_npu` tracking issue in a comment on the test, or narrow the GRU case so it does not claim coverage it does not have. This also does not block pytorch/pytorch CI (no NPU CI exists), which is precisely why it will go unnoticed.
> 
> **4. On a CPU-only host these four tests now vanish rather than skip.** `test/export/test_passes.py:1450`
> 
> `instantiate_device_type_tests` unconditionally does `del scope[generic_test_class.__name__]` (`common_device_type.py:1108`) and then generates classes only for available bases (`:1114`). With `except_for="cpu"` and no accelerator, zero test classes are produced. Previously `@unittest.skipIf(not TEST_CUDA, "requires cuda")` surfaced four explicit skips in the CI report. This is a standard tradeoff of the framework rather than a bug, but it does mean the loss of coverage on a misconfigured runner reads as green rather than as skipped — worth being deliberate about.
> 
> ### Code Quality
> **5. `hw_classification = HardwareClassification.GENERIC` on `TestPasses` is unrelated to this PR.** `test/export/test_passes.py:378`
> 
> `TestPasses` is not passed to `instantiate_device_type_tests` and is otherwise untouched by this change. The classification is defensible on its own (it makes the class visible under `--hw-classification GENERIC` filtering), but it is undescribed in the PR body and expands the blast radius of a test-refactor PR. Either mention it in the description or split it out.
> 
> Note for reviewers: `HardwareClassification` does not exist at this branch's merge base (`da9c79b0`) — it was added to `torch/testing/_internal/common_utils.py:127` on `main` afterwards. The file therefore cannot import on the branch as checked out; it only resolves against current `main`. CI tests the merge commit so this should be green, but a rebase would make local verification possible.
> 
> **6. `test_move_device_submod` mixes type-only and indexed device without explanation.** `test/export/test_passes.py:1358-1368`
> 
> `torch.autocast(device_type=device_type)` takes the type, `x.to(device=device)` takes the indexed form, and `move_to_device_pass(ep, device_type)` takes the type again. Each choice is individually correct, but the three-way split within eight lines is non-obvious to a future reader. One short comment noting that `autocast` and the pass target a device _type_ while the tensor move targets a concrete device would carry its weight here.
> 
> ### Verified as correct (no action needed)
> Recorded only because they were plausible failure modes I checked and ruled out, so they need not be re-investigated:
> 
> * `except_for="cpu"` as a bare string is handled — `common_device_type.py:968-972` explicitly branches on `isinstance(except_for, str)`.
> * `assertEqual` on two `torch.device` objects compares type _and_ index (`c10/core/Device.h:50-51`); `.to("cuda:0").device == torch.device("cuda:0")` holds.
> * `{"cpu": str(target_device)}` at `:1442` genuinely exercises the dict-remap branch: a CPU device has `index == -1`, so `Device::str()` (`c10/core/Device.cpp:151-158`) yields `"cpu"` and the key matches. A miss would fail loudly at `:1447`, not pass vacuously.
> * Baking the indexed `device` into the module at `:1341` while calling the pass with `device_type` preserves the pre-PR semantics and remains correct for a nonzero primary device index, since the pass overwrites the burnt-in device with the type-only form.
> * Class-level `@skipIfTorchDynamo` / `@unittest.skipIf` survive instantiation — they set `__unittest_skip__` as a class attribute (`common_utils.py:1938-1940`) and the generated class inherits it via `type(name, (base, generic_test_class), {})` (`common_device_type.py:1125`).
> * `hw_classification` is likewise inherited by the generated device classes and is complementary to `except_for`, not redundant with it.
> 
> ### Recommendation
> **Request Changes**
> 
> Finding 1 is a concrete regression in test maintainability with a cheap, established fix already used elsewhere in the repo. Finding 2 asks only for evidence that the device coverage this PR advertises actually passes — given two goldens and an on-device GRU, an NPU-only test run is not enough to land XPU and CUDA enablement. Finding 3 should be resolved by a tracking-issue reference rather than by weakening the test.
> 
> | Branch: `npu-export-device-agnostic-passes`

Rebased onto current main and pushed the follow-ups from the review.

Goldens: static 'cuda' literals plus .replace("cuda", device_type) so EXPECTTEST_ACCEPT can still rewrite the source. No GPU_TYPE.
Dropped allow_xpu=True. The pre-PR tests were @skipIf(not TEST_CUDA) only; nobody asked to add XPU. CUDA coverage is what instantiate enables by default.
GRU on NPU: this is a torch_npu backend gap (DynamicGRUV2), not a move_to_device_pass bug. The case stays on the accelerator class (no per-device skip). Tracking: https://gitcode.com/Ascend/pytorch/issues/3886
TestPasses / GENERIC: only so --hw-classification GENERIC still sees the remaining CPU tests. Unrelated to instantiate; left in this PR and called out in the description.
CPU-only hosts: four skipIf(not TEST_CUDA) skips become “no generated class”. That is the usual except_for="cpu" tradeoff.
test_move_device_submod: autocast and the pass take a device type; .to(device=device) takes the injected device. No code change.

**Comment by SHIHAIFENG15** [comment]
@can-gaa-hou  Claude's review comments have been addressed; the changes are as described in the reply above.

**Comment by SHIHAIFENG15** [comment]
Hi @ezyang, this PR has already received a review and has been updated to
address the feedback. Could you please help land it or route it to the
appropriate maintainer? Thanks!

**Comment by ezyang** [comment]
nope @fffrog is the guy you need
